include/aspect/particle/property/interface.h

```diff
@@ -0,0 +1,355 @@
+
+#include <deal.II/base/std_cxx1x/shared_ptr.h>
+
+namespace aspect
+{
+  namespace Particle
+  {
+    namespace Property
+    {
+
+      enum UpdateTimeFlags
+      {
+        /**
+         * Never update the initially set properties.
+         */
+        update_never,
+        /**
+         * Update the tracer properties before outputting them. This is
+         * sufficient for all tracer properties that depend on the current
+         * solution, like current velocity or pressure.
+         */
+        update_output_step,
+        /**
+         * Update the tracer properties every timestep. This is only necessary
+         * if the properties at the output time depend on some sort of time
+         * integration of solution properties.
+         */
+        update_time_step
+      };
+
+      /**
+        * Interface provides an example of how to extend the Particle
+        * class to include related particle data. This allows users to attach
+        * scalars/vectors/tensors/etc to particles and ensure they are
+        * transmitted correctly over MPI and written to output files.
+        */
+      template <int dim>
+      class Interface
+      {
+        public:
+
+          /**
+           * Initialization function. This function is called once at the
+           * beginning of the program after parse_parameters is run.
+           */
+          virtual
+          void
+          initialize ();
+
+          /**
+           * Initialization function. This function is called once at the
+           * creation of every particle to set up it's properties.
+           */
+          virtual
+          void
+          initialize_particle (std::vector<double> &data,
+                               const Point<dim> &position,
+                               const Vector<double> &solution,
+                               const std::vector<Tensor<1,dim> > &gradients);
+
+          /**
+           * Update function. This function is called every timestep for
+           * every particle to update it's properties. It is obvious that
+           * this function is called a lot, so its code should be efficient.
+           */
+          virtual
+          void
+          update_particle (unsigned int &data_position,
+                           std::vector<double> &particle_properties,
+                           const Point<dim> &position,
+                           const Vector<double> &solution,
+                           const std::vector<Tensor<1,dim> > &gradients);
+
+          /**
+           * Returns a bool, which is false in the default implementation,
+           * telling the property manager that no update is needed. Every
+           * plugin that implements this function should return true. This
+           * saves considerable computation time in cases, when no plugin needs
+           * to update tracer properties over time.
+           */
+          virtual
+          UpdateTimeFlags
+          need_update ();
```

Comment:
The documentation talks about true/false, but the function actually returns an enum.